### ISSUE TITLE
Records: separate field names from binders

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#5361]
In theoretical treatments of record declarations, e.g. http://www-sop.inria.fr/oasis/DTP00/Proceedings/pollack.ps , the label names (which are not subject to alpha-renaming) are distinguished from their role as binders (which would be subject to alpha-renaming).  We could support this e.g. by the following syntax:
```agda
record Sigma (A : Set) (B : A -> Set) : Set where
  field
    fst@a : A
    snd   : B a

pair : A -> B -> Sigma A \ _ -> B
pair a b .fst = a
pair a b .snd = b

record Congruence (A B : Set) : Set where
   field
     apply@f : A -> B
     cong    : {a a' : A} -> a == a' -> f a == f a' 

idCong : Congruence A A
idCong .apply x = x
idCong .cong  p = p
```
Note how `apply` would be awkward to read in the definition of `Congruence.cong`, and `.f` would be awkward to read in the first clause for `idCong`.

Other syntax idea: use `=` instead of `@` (which is also used for attributes that can prefix field names):
```agda
  field
    fst = a : A
    snd     : B a
```
This syntax would align with record expressions `record { fst = a; snd = b }`.

(Issue triggered by https://github.com/agda/agda-stdlib/issues/1493.)

**Comment by jespercockx** [comment]
I think being able to rename the fields in this way would be a very nice feature to have. For the bikeshedding part, I'm in favour of your second proposal, since it is similar to the existing syntax we have for renaming function arguments `{x = y : A}`, as well as how shadowed context variables are displayed in the interactive mode.

**Comment by JacquesCarette** [comment]
So which one is the label and which one is the binder? Which one is supposed to be used when?

This seems like a good idea. I'm extremely interested in the meta-theory of such a feature, which is why I'm asking for more details.

While I understand that `=` is consistent with current uses, `=` is already massively overloaded. Plus `=` is symmetric while all these uses are fundamentally oriented. I think considering an oriented notation would be worthwhile.

**Comment by UlfNorell** [comment]
+1 for `fst = a`.

This syntax is currently used for
- renamed implicit function arguments
- shadowed context variables
- record expressions and patterns
- giving and matching on implicit arguments by name

In all of these cases it's `label = value/pattern/binding` so I don't see how adding one more consistent use case would add any (additional) confusion.

**Comment by andreasabel** [comment]
@JacquesCarette: 
> So which one is the label and which one is the binder? Which one is supposed to be used when?

I was asking myself this question for the suggestion `x@y`, but for `x = y` I can hold on to the analogy of record expressions as a handrail.

An open question is how to scale this to record extension:
```agda
record Semigroup : Set1 where
  field
    Carrier = A   : Set
    op      = _*_ : A -> A -> A
    assoc         : {a b c : A} -> ((a * b) * c) == (a * (b * c))

record Monoid : Set2 where
  field
    semigroup : Semigroup
  module S = Semigroup semigroup 
  open S public
  open S using () renaming (Carrier to A; op to _*_)
  field
    unit = e   : A
    unit-left  : {a : A} -> (e * a) == a
    unit-right : {a : A} -> (a * e) == a
```

**Comment by andreasabel** [comment]
Actually, we can already simulate the feature using `private` declaration, for the price of extra noise:
```agda
open import Agda.Builtin.Equality

record Magma : Set1 where
  field
    Carrier : Set
  private A = Carrier
  field
    op      : (x y : A) → A

record Semigroup : Set2 where
  field
    magma : Magma
  open Magma magma public
  open Magma magma renaming (Carrier to A; op to infixl 5 _*_)
  field
    assoc : ∀ {a b c : A} → (a * b) * c ≡ a * (b * c)

record Monoid : Set3 where
  field
    semigroup : Semigroup
  open Semigroup semigroup public
  open Semigroup semigroup renaming (Carrier to A; op to infixl 5 _*_)
  field
    unit  : A
  private e = unit
  field
    unit-left  : ∀{a} → e * a ≡ a
    unit-right : ∀{a} → a * e ≡ a

infixl 5 _*_

postulate
  M          : Set
  _*_        : (x y : M) → M
  assoc      : ∀ {x y z : M} → (x * y) * z ≡ x * (y * z)
  e          : M
  unit-left  : ∀ {x : M} → e * x ≡ x
  unit-right : ∀ {x : M} → x * e ≡ x

m : Monoid
m = record
  { semigroup  = record
   { magma     = record
     { Carrier = M
     ; op      = _*_
     }
   ; assoc     = assoc
   }
  ; unit       = e
  ; unit-left  = unit-left
  ; unit-right = unit-right
  }
```

Well, mostly.  Since the `private` declarations are not bound variables, they can clash with things of the same name already in scope, unless we fix #3801.

**Comment by JacquesCarette** [comment]
In your first example (second-last message), why do you need
```agda
  open S using () renaming (Carrier to A; op to _*_)
```
since you had already "done that" in the declaration of `Semigroup`. This is why I asked what you mean by 'label' and 'binder', as I honestly don't know. [Of course, I know the English and intuitive PL meanings, but not the technical meaning you/Agda attach to them. I guess I should go re-read that paper...]

Your simulation also has another effect: both the old and new names are visible. So this contributes serious pollution to the namespace?

Let me make a guess (from your usage):
- *binder* = name of a field that's visible externally to the record, especially when you want to set it
- *label* = name of a field you can use to refer to the value of that field internally to the record definition.

**Comment by andreasabel** [comment]
@JacquesCarette 
> In your first example (second-last message), why do you need
> 
> ```agda
>   open S using () renaming (Carrier to A; op to _*_)
> ```

Yeah, this is an open question, if the binders could survive their use in the record declaration and be exported somehow.  But that would contradict their spirit, they should be local (bound variables).

C.f. `data` declarations, e.g.
```agda
data Sigma (A : Set) (B : A -> Set) : Set where
  pair : (x : A) -> B x -> Sigma A B
```
Here `x` is just a bound variable.

>     * _binder_ = name of a field that's visible externally to the record, especially when you want to set it
> 
>     * _label_ = name of a field you can use to refer to the value of that field internally to the record definition.

Actually, I meant to assign the meaning the other way round...

**Comment by JacquesCarette** [comment]
That's the usefulness of expanding out one's informal meaning...

So there seems to be two concepts here, "the field name", which is visible externally, and "the name of the value of that field", which is not.

It seems that the underlying reasons to want this are: 1) usability and 2) wishing to not pollute the namespace too much. So `Carrier` is a good enough name externally, but taking over `A` would be too much. But `Carrier` is rather long, we it would be nice to have a shorter name "in practice".

It feels like you want 'local aliases'.

Radical suggestion:
```agda
record Semigroup : Set1 where
  field
    Carrier = A   : Set
    op      = _*_ : A -> A -> A
    assoc         : {a b c : A} -> ((a * b) * c) == (a * (b * c))
```
would desugar to
```agda
record Semigroup : Set1 where
  field
    Carrier  : Set
  alias _ Carrier = A
  field
    op      : A -> A -> A
  alias _ op = _*_
  field
    assoc         : {a b c : A} -> ((a * b) * c) == (a * (b * c))
```
why the extra complication?  To also allow:
```agda
record Semigroup : Set1 where
  field
    Carrier = A   : Set
    op      = _*_ : A -> A -> A
    assoc         : {a b c : A} -> ((a * b) * c) == (a * (b * c))
  alias S2 {op = _+_} -- automatically inherits Carrier = A  ?
```
Then you could say
```agda
open S renaming alias
```
or
```agda
open S renaming alias S2
```

The name (`alias`) and syntax that are used above should not be considered a serious proposal. The only serious part of my proposal is the idea of associating a set of named aliases to a record. One can think of it as 'associated renames'. It sure would be nice to be able to use them externally too, i.e. `record @alias S2 { ... }` for example.
